File: app/src/main/java/ch/ralena/natibo/adapter/CourseListAdapter.java
package ch.ralena.natibo.adapter;

import android.support.annotation.NonNull;
import android.support.v7.widget.RecyclerView;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ImageView;
import android.widget.TextView;

import ch.ralena.natibo.R;
import ch.ralena.natibo.object.Course;
import io.reactivex.subjects.PublishSubject;
import io.realm.RealmResults;

public class CourseListAdapter extends RecyclerView.Adapter<CourseListAdapter.ViewHolder> {

	PublishSubject<Course> courseSubject = PublishSubject.create();

	public PublishSubject<Course> asObservable() {
		return courseSubject;
	}

	private RealmResults<Course> courses;

	public CourseListAdapter(RealmResults<Course> courses) {
		this.courses = courses;
	}

	@NonNull
	@Override
	public ViewHolder onCreateViewHolder(@NonNull ViewGroup parent, int viewType) {
		View view;
		view = LayoutInflater.from(parent.getContext()).inflate(R.layout.item_course_list, parent, false);
		return new ViewHolder(view);
	}

	@Override
	public void onBindViewHolder(@NonNull ViewHolder holder, int position) {
		if (position < getItemCount())
			holder.bindView(courses.get(position));
	}

	@Override
	public int getItemCount() {
		return courses.size();
	}

	class ViewHolder extends RecyclerView.ViewHolder {
		private View view;
		private TextView languageName;
		private TextView courseTitle;
		private TextView numReps;
		private ImageView flagImage;
		private Course course;

		ViewHolder(View view) {
			super(view);
			this.view = view;
			languageName = view.findViewById(R.id.languageLabel);
			courseTitle = view.findViewById(R.id.courseTitleLabel);
			numReps= view.findViewById(R.id.numRepsLabel);
			flagImage = view.findViewById(R.id.flagImageView);
			this.view.setOnClickListener(v -> courseSubject.onNext(course));
		}

		void bindView(Course course) {
			this.course = course;
			languageName.setText(course.getTargetLanguage().getLongName());
			courseTitle.setText(course.getTitle());
			numReps.setText(String.format("%d reps", course.getTotalReps()));
			flagImage.setImageResource(course.getTargetLanguage().getLanguageType().getDrawable());
		}
	}
}

File: app/src/main/java/ch/ralena/natibo/ui/course/list/adapter/CourseListAdapter.kt
package ch.ralena.natibo.ui.course.list.adapter

import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.ImageView
import android.widget.TextView
import androidx.recyclerview.widget.RecyclerView
import ch.ralena.natibo.R
import ch.ralena.natibo.data.room.`object`.CourseRoom
import ch.ralena.natibo.data.room.`object`.LanguageRoom
import ch.ralena.natibo.ui.base.BaseRecyclerAdapter

class CourseListAdapter(
	private val courses: ArrayList<CourseRoom>,
	private val languages: ArrayList<LanguageRoom>
) :
	BaseRecyclerAdapter<CourseListAdapter.ViewHolder, CourseListAdapter.Listener>() {
	interface Listener {
		fun onCourseClicked(course: CourseRoom)
	}

	override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): ViewHolder {
		val view = LayoutInflater.from(parent.context)
			.inflate(R.layout.item_course_list, parent, false)
		return ViewHolder(view)
	}

	override fun onBindViewHolder(holder: ViewHolder, position: Int) {
		if (position < itemCount) holder.bindView(courses[position])
	}

	override fun getItemCount() = courses.size

	fun loadData(courses: List<CourseRoom>, languages: List<LanguageRoom>) {
		this.courses.clear()
		this.courses.addAll(courses)
		this.languages.clear()
		this.languages.addAll(languages)
		notifyDataSetChanged()
	}

	inner class ViewHolder(view: View) : RecyclerView.ViewHolder(view) {
		private val languageName = view.findViewById<TextView>(R.id.languageLabel)
		private val courseTitle = view.findViewById<TextView>(R.id.courseTitleLabel)
		private val numReps = view.findViewById<TextView>(R.id.numRepsLabel)
		private val flagImage = view.findViewById<ImageView>(R.id.flag_image)

		private lateinit var course: CourseRoom

		init {
			view.setOnClickListener {
				listeners.forEach { it.onCourseClicked(course) }
			}
		}

		fun bindView(course: CourseRoom) {
			this.course = course
			courseTitle.text = course.title
			// TODO: Use String resource instead
//			numReps.text = String.format("%d reps", course.totalReps)
			val language = languages.find { it.id == course.targetLanguageId }
			language?.let {
				languageName.text = it.name
				flagImage.setImageResource(it.flagDrawable)
			}
		}
	}
}